src/coord/src/coord.rs

```diff
@@ -5217,54 +5284,70 @@ mod timeline {
+    impl<T: super::CoordTimestamp> TimestampOracle<T> {
+        /// Create a new timeline, starting at the indicated time. `next` generates
+        /// new timestamps when invoked. The timestamps have no requirements, and can
+        /// retreat from previous invocations.
+        pub fn new<F>(initially: T, next: F) -> Self
+        where
+            F: Fn() -> T + 'static,
+        {
             Self {
-                state: TimestampOracleState::Writing(initially),
+                state: TimestampOracleState::Writing(initially.clone()),
                 advance_to: Some(initially),
+                next: Box::new(next),
             }
         }
 
         /// Acquire a new timestamp for writing.
         ///
-        /// This timestamp will be strictly greater than all prior values of `self.read_ts()`,
-        /// and less than or equal to all subsequent values of `self.read_ts()`.
-        pub fn write_ts(&mut self) -> mz_repr::Timestamp {
-            match self.state {
-                TimestampOracleState::Writing(ts) => ts,
+        /// This timestamp will be strictly greater than all prior values of
+        /// `self.read_ts()`, and less than or equal to all subsequent values of
+        /// `self.read_ts()`. If a timestamp meeting that criteria was not able to be
+        /// produced, return an `Err` with the current timestamp.
+        pub fn write_ts(&mut self) -> Result<T, T> {
+            match &self.state {
+                TimestampOracleState::Writing(ts) => Ok(ts.clone()),
                 TimestampOracleState::Reading(ts) => {
-                    self.state = TimestampOracleState::Writing(ts + 1);
-                    ts + 1
+                    let next = (self.next)();
+                    if next.less_equal(&ts) {
+                        Err(ts.clone())
+                    } else {
+                        self.state = TimestampOracleState::Writing(next.clone());
+                        Ok(next)
+                    }
                 }
             }
         }
         /// Acquire a new timestamp for reading.
         ///
         /// This timestamp will be greater or equal to all prior values of `self.write_ts()`,
         /// and strictly less than all subsequent values of `self.write_ts()`.
-        pub fn read_ts(&mut self) -> mz_repr::Timestamp {
-            match self.state {
-                TimestampOracleState::Reading(ts) => ts,
+        pub fn read_ts(&mut self) -> T {
+            match &self.state {
+                TimestampOracleState::Reading(ts) => ts.clone(),
                 TimestampOracleState::Writing(ts) => {
-                    self.state = TimestampOracleState::Reading(ts);
-                    self.advance_to = Some(ts + 1);
+                    // Avoid rust borrow complaint.
+                    let ts = ts.clone();
+                    self.state = TimestampOracleState::Reading(ts.clone());
+                    self.advance_to = Some(ts.step_forward());
```

Comment:
`advance_to` isn't the timestamp, it's an advertisement to coord that things can compact up to (and not including) the timestamp. So for a read at `ts`, we can tell coord it can compact up to `ts+1`.

Reply:
Ah, I wasn't paying close enough attention to what `advance` means in these contexts; ty!